onnxoptimizer/passes/eliminate_nop_with_unit.cc

```diff
@@ -0,0 +1,304 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#include "eliminate_nop_with_unit.h"
+#include "onnx/defs/tensor_util.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace optimization;
+
+namespace {
+
+#define PROTO_DTYPE_TO_CPP_DTYPE_LIST(_)             \
+  _(TensorProto_DataType_FLOAT, float, 1, 0)         \
+  _(TensorProto_DataType_INT32, int32_t, 1, 0)       \
+  _(TensorProto_DataType_INT64, int64_t, 1, 0)       \
+  _(TensorProto_DataType_DOUBLE, double, 1, 0)       \
+  _(TensorProto_DataType_FLOAT16, int32_t, 15360, 0) \
+  _(TensorProto_DataType_UINT8, int32_t, 1, 0)       \
+  _(TensorProto_DataType_INT8, int32_t, 1, 0)        \
+  _(TensorProto_DataType_UINT16, int32_t, 1, 0)      \
+  _(TensorProto_DataType_INT16, int32_t, 1, 0)       \
+  _(TensorProto_DataType_UINT32, uint64_t, 1, 0)     \
+  _(TensorProto_DataType_UINT64, uint64_t, 1, 0)
+
+enum Unit {
+  ZERO,
+  ONE,
+};
+
+template <TensorProto_DataType, Unit>
+struct IdentityUnitTraits;
+
+#define IDENTITY_UNIT_TRAITS(pb_dtype, cpp_dtype, one, zero)                 \
+  template <>                                                                \
+  struct IdentityUnitTraits<pb_dtype, ZERO> {                                \
+    using cpp_type = cpp_dtype;                                              \
+    static bool isAllValue(const std::vector<cpp_type>& data) {              \
+      return std::all_of(data.cbegin(), data.cend(), [](const cpp_type& v) { \
+        return v == cpp_type{zero};                                          \
+      });                                                                    \
+    }                                                                        \
+  };                                                                         \
+  template <>                                                                \
+  struct IdentityUnitTraits<pb_dtype, ONE> {                                 \
+    using cpp_type = cpp_dtype;                                              \
+    static bool isAllValue(const std::vector<cpp_type>& data) {              \
+      return std::all_of(data.cbegin(), data.cend(), [](const cpp_type& v) { \
+        return v == cpp_type{one};                                           \
+      });                                                                    \
+    }                                                                        \
+  };
```

Comment:
Great work.

Some refactoring is needed here. There is no need to provide the ZERO and the ONE simultaneously since an operation has only one identity element, and it can not scale when taking more operations-identity pairs into account.

It is better to maintain a one-to-one map from operation (e.g. kAdd) to the identity element (e.g. 0).

```cpp
template<uint32_t kind, class T>
struct IdentityElement;

template<class T>
struct IdentityElement<kAdd, T> {
  T get() { return static_cast<T>(0); }
};
```

So that `NODE_KIND_TRAITS_SUPPORT_COMMUTATIVE_LAW` can accept only the operation as argument and get the identity element by IdentityElement<kOp, T>::get()